[BUG] goroutine leak in `init()` of `v2/cmd/formatter` ?

### Description

Trying to build something that depends on `github.com/docker/compose/v2`, then ran into this issue, because all of our tests checks for goroutine leaks at the end of running.

### Steps To Reproduce

Run following test


```go
package main

import (
	"testing"

	_ "github.com/docker/compose/v2/cmd/formatter"
	"go.uber.org/goleak"
)

func TestLeak(t *testing.T) {
	defer goleak.VerifyNone(t)
}
```

Output:


```shell
$ go test
--- FAIL: TestLeak (0.45s)
    main_test.go:12: found unexpected goroutines:
        [Goroutine 19 in state chan send, with github.com/docker/compose/v2/cmd/formatter.init.0.func1 on top of the stack:
        goroutine 19 [chan send]:
        github.com/docker/compose/v2/cmd/formatter.init.0.func1()
                /home/wxh/go/pkg/mod/github.com/docker/compose/v2@v2.14.2/cmd/formatter/colors.go:120 +0x1de
        created by github.com/docker/compose/v2/cmd/formatter.init.0
                /home/wxh/go/pkg/mod/github.com/docker/compose/v2@v2.14.2/cmd/formatter/colors.go:104 +0x8b
        ]
FAIL
exit status 1
FAIL    tt      0.451s
```


This is because there is a non-terminating goroutine in the `init()`:

```go
	go func() {
		i := 0
		rainbow := []colorFunc{
			colors["cyan"],
			colors["yellow"],
			colors["green"],
			colors["magenta"],
			colors["blue"],
			colors["intense_cyan"],
			colors["intense_yellow"],
			colors["intense_green"],
			colors["intense_magenta"],
			colors["intense_blue"],
		}

		for {
			loop <- rainbow[i]
			i = (i + 1) % len(rainbow)
		}
	}()
```

### Workaround

Not a best practice to bypass questionable func, but...

```go
	defer goleak.VerifyNone(t, goleak.IgnoreTopFunction("github.com/docker/compose/v2/cmd/formatter.init.0.func1"))
```

### Compose Version

```Text
https://github.com/docker/compose/blob/v2.14.2/cmd/formatter/colors.go
```


### Docker Environment

_No response_

### Anything else?

_No response_

--- a/cmd/formatter/colors.go
+++ b/cmd/formatter/colors.go
@@ -99,6 +99,7 @@ func cleanInfiniteGoroutine() {
 }
 
 func init() {
+	defer cleanInfiniteGoroutine()
 	colors := map[string]colorFunc{}
 	for i, name := range names {
 		colors[name] = makeColorFunc(strconv.Itoa(30 + i))
@@ -131,5 +132,4 @@ func init() {
 		}
 	}()
 
-	defer cleanInfiniteGoroutine()
 }

--- a/cmd/formatter/colors.go
+++ b/cmd/formatter/colors.go
@@ -92,6 +92,11 @@ func rainbowColor() colorFunc {
 }
 
 var loop = make(chan colorFunc)
+var quitChan = make(chan bool)
+
+func cleanInfiniteGoroutine() {
+	quitChan <- true
+}
 
 func init() {
 	colors := map[string]colorFunc{}
@@ -116,8 +121,15 @@ func init() {
 		}
 
 		for {
-			loop <- rainbow[i]
-			i = (i + 1) % len(rainbow)
+			select {
+			case <-quitChan:
+				return
+			default:
+				loop <- rainbow[i]
+				i = (i + 1) % len(rainbow)
+			}
 		}
 	}()
+
+	defer cleanInfiniteGoroutine()
 }

--- a/cmd/formatter/formatter_test.go
+++ b/cmd/formatter/formatter_test.go
@@ -75,5 +75,5 @@ func TestPrint(t *testing.T) {
 
 // Test the absence of unexpected goroutines.
 func TestColorsGoroutinesLeak(t *testing.T) {
-	defer goleak.VerifyNone(t)
+	goleak.VerifyNone(t)
 }

--- a/cmd/formatter/formatter_test.go
+++ b/cmd/formatter/formatter_test.go
@@ -22,6 +22,7 @@ import (
 	"io"
 	"testing"
 
+	"go.uber.org/goleak"
 	"gotest.tools/v3/assert"
 )
 
@@ -70,4 +71,9 @@ func TestPrint(t *testing.T) {
 	assert.Equal(t, json, `{"Name":"myName1","Status":"myStatus1"}
 {"Name":"myName2","Status":"myStatus2"}
 `)
+}
+
+// Test the absence of unexpected goroutines.
+func TestColorsGoroutinesLeak(t *testing.T) {
+	defer goleak.VerifyNone(t)
 }

--- a/go.mod
+++ b/go.mod
@@ -150,6 +150,8 @@ require (
 	sigs.k8s.io/yaml v1.2.0 // indirect
 )
 
+require go.uber.org/goleak v1.1.12
+
 replace (
 	// Override for e2e tests
 	github.com/cucumber/godog => github.com/laurazard/godog v0.0.0-20220922095256-4c4b17abdae7

